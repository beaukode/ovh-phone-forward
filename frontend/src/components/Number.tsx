import React from "react";
import {
  ListItem,
  ListItemText,
  ListItemSecondaryAction,
  Switch,
  Box,
  makeStyles,
  Theme,
  createStyles,
} from "@material-ui/core";
import { PhoneForwarded as PhoneForwardedIcon } from "@material-ui/icons";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    secondary: {
      display: "flex",
      alignItems: "center",
    },
  })
);

type ChangeCallback = (number: string, forward: string | null) => void;

interface Props {
  number: string;
  forward: string | null;
  onChange?: ChangeCallback;
}

const Number: React.FC<Props> = ({ number, forward, onChange }) => {
  const classes = useStyles();

  const handleChange = (
    event: React.ChangeEvent<HTMLInputElement>,
    checked: boolean
  ) => {
    if (onChange) {
      onChange(number, checked ? null : forward);
    }
  };

  const id = "number-" + number;
  const fwNode = forward ? (
    <>
      <PhoneForwardedIcon />
      <Box component="span" flexGrow={1} margin={1}>
        {forward}
      </Box>
    </>
  ) : null;
  return (
    <ListItem>
      <ListItemText
        primary={number}
        primaryTypographyProps={{ id }}
        secondary={fwNode}
        secondaryTypographyProps={{ className: classes.secondary }}
      />
      <ListItemSecondaryAction>
        <Switch
          data-testid="switch"
          edge="end"
          inputProps={{ "aria-labelledby": id }}
          checked={Boolean(forward)}
          onChange={handleChange}
        />
      </ListItemSecondaryAction>
    </ListItem>
  );
};

export default Number;
