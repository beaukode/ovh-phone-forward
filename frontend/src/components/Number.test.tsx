import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Number from "./Number";

describe("<Number />", () => {
  test("Render", () => {
    const { container, getByTestId } = render(
      <Number number="number-1" forward={null} />
    );
    expect(container).toMatchSnapshot();
    expect(getByTestId("switch")).toBeInTheDocument();
  });

  test("Render forwarded", () => {
    const { container, getByTestId } = render(
      <Number number="number-1" forward="fwnumber-2" />
    );
    expect(container).toMatchSnapshot();
    expect(getByTestId("switch")).toBeInTheDocument();
  });

  test("Change", () => {
    const callback = jest.fn();
    const { getByRole } = render(
      <Number number="number-1" forward={null} onChange={callback} />
    );
    fireEvent.click(getByRole("checkbox"));
    expect(callback).toBeCalledTimes(1);
    expect(callback).toBeCalledWith("number-1", null);
  });

  test("Change forwarded", () => {
    const callback = jest.fn();
    const { getByRole } = render(
      <Number number="number-1" forward="fwnumber-2" onChange={callback} />
    );
    fireEvent.click(getByRole("checkbox"));
    expect(callback).toBeCalledTimes(1);
    expect(callback).toBeCalledWith("number-1", "fwnumber-2");
  });
});
