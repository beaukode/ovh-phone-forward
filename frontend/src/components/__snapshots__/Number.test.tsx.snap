// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Number /> Render 1`] = `
<div>
  <li
    class="MuiListItem-container"
  >
    <div
      class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
    >
      <div
        class="MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          id="number-number-1"
        >
          number-1
        </span>
      </div>
    </div>
    <div
      class="MuiListItemSecondaryAction-root"
    >
      <span
        class="MuiSwitch-root MuiSwitch-edgeEnd"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-62 MuiSwitch-switchBase MuiSwitch-colorSecondary"
          data-testid="switch"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              aria-labelledby="number-number-1"
              class="PrivateSwitchBase-input-65 MuiSwitch-input"
              type="checkbox"
              value=""
            />
            <span
              class="MuiSwitch-thumb"
            />
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiSwitch-track"
        />
      </span>
    </div>
  </li>
</div>
`;

exports[`<Number /> Render forwarded 1`] = `
<div>
  <li
    class="MuiListItem-container"
  >
    <div
      class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
    >
      <div
        class="MuiListItemText-root MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          id="number-number-1"
        >
          number-1
        </span>
        <p
          class="MuiTypography-root makeStyles-secondary-88 MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 11l5-5-5-5v3h-4v4h4v3zm2 4.5c-1.25 0-2.45-.2-3.57-.57-.35-.11-.74-.03-1.02.24l-2.2 2.2c-2.83-1.44-5.15-3.75-6.59-6.59l2.2-2.21c.28-.26.36-.65.25-1C8.7 6.45 8.5 5.25 8.5 4c0-.55-.45-1-1-1H4c-.55 0-1 .45-1 1 0 9.39 7.61 17 17 17 .55 0 1-.45 1-1v-3.5c0-.55-.45-1-1-1z"
            />
          </svg>
          <span
            class="MuiBox-root MuiBox-root-146"
          >
            fwnumber-2
          </span>
        </p>
      </div>
    </div>
    <div
      class="MuiListItemSecondaryAction-root"
    >
      <span
        class="MuiSwitch-root MuiSwitch-edgeEnd"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-160 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-161 Mui-checked"
          data-testid="switch"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              aria-labelledby="number-number-1"
              checked=""
              class="PrivateSwitchBase-input-163 MuiSwitch-input"
              type="checkbox"
              value=""
            />
            <span
              class="MuiSwitch-thumb"
            />
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiSwitch-track"
        />
      </span>
    </div>
  </li>
</div>
`;
